gepetto-viewer: init at 5.1.0

nim65s opened the pull request:
## Description of changes

This add the gepetto-viewer GUI, an openscenegraph & Qt application, with its python interface and CORBA middleware:
https://github.com/gepetto/gepetto-viewer

Here is an example on how to use it to display the talos robot:
```nix
{
  description = "Test gepetto-viewer";
  inputs.nixpkgs.url = "github:nim65s/nixpkgs/gepetto-viewer";
  outputs =
    { nixpkgs, ... }:
    let
      system = "x86_64-linux";
      pkgs = nixpkgs.legacyPackages.${system};
    in
    {
      devShells.${system}.default = pkgs.mkShell {
        packages = [
          (pkgs.python3.withPackages (p: [ 
            p.example-robot-data 
            p.gepetto-gui 
          ]))
        ];
      };
    };
}
```
And in that shell:
```
gepetto-gui & python -m example_robot_data talos
```

This should display (after clicking on "zoom to fit" and moving around a bit):
![screenshot](https://github.com/NixOS/nixpkgs/assets/131929/7b470a8b-49dc-43ac-b0dd-5ffe144d5359)


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nim65s committed:
osgqt: init at 3.5.7

nim65s committed:
omniorbpy: move to python-modules

nim65s committed:
omniorb: add in python3Packages

nim65s committed:
gepetto-viewer: init at 5.1.0

nim65s committed:
qgv: init at 1.3.5

wegank commented:
Looks good on Linux, though `gepetto-viewer` doesn't build on darwin.

nim65s commented:
thanks for the test, I'll investigate this week

nim65s commented:
Ok, so deactivating a couple optional plugins on darwin fix the build, but I have no idea how to start the GUI. running `gepetto-gui` in that development shell immediately raise `Killed: 9`.

wegank commented:
I guess `darwin.autoSignDarwinBinariesHook` is needed somewhere. Does the issue occur on x86_64-darwin?

nim65s committed:
python-qt: fix link path

fix runtime error:
```
dyld[23235]: Library not loaded: libPythonQt-Qt5-Python3.12.3.dylib
  Referenced from: <C9EF5524-29A1-3E24-ABD4-DDEC31CF4B47> /nix/store/gnsk805im3lfr3zglmdap5p6j3zjpn05-python-qt-3.5.4/lib/libPythonQt_QtAll-Qt5-Python3.12.3.2.0.dylib
  Reason: tried: 'libPythonQt-Qt5-Python3.12.3.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibPythonQt-Qt5-Python3.12.3.dylib' (no such file), 'libPythonQt-Qt5-Python3.12.3.dylib' (no such file), '/usr/local/lib/libPythonQt-Qt5-Python3.12.3.dylib' (no such file), '/usr/lib/libPythonQt-Qt5-Python3.12.3.dylib' (no such file, not in dyld cache), '/Users/gsaurel/local/NixOS/nixpkgs/libPythonQt-Qt5-Python3.12.3.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/gsaurel/local/NixOS/nixpkgs/libPythonQt-Qt5-Python3.12.3.dylib' (no such file), '/Users/gsaurel/local/NixOS/nixpkgs/libPythonQt-Qt5-Python3.12.3.dylib' (no such file), '/usr/local/lib/libPythonQt-Qt5-Python3.12.3.dylib' (no such file), '/usr/lib/libPythonQt-Qt5-Python3.12.3.dylib' (no such file, not in dyld cache)
```

nim65s commented:
Thanks ! I can now load and display meshes on aarch64-darwin, but this require #325214 to load those meshes from example-robot-data. Here is a branch to test that: https://github.com/nim65s/nixpkgs/tree/pin310%2Bgepetto-viewer

nim65s commented:
Result of `nixpkgs-review pr 324018` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>18 packages built:</summary>
  <ul>
    <li>gepetto-viewer</li>
    <li>osgqt</li>
    <li>python-qt</li>
    <li>python311Packages.crocoddyl</li>
    <li>python311Packages.example-robot-data</li>
    <li>python311Packages.omniorb</li>
    <li>python311Packages.omniorbpy</li>
    <li>python311Packages.omniorbpy.dev</li>
    <li>python312Packages.crocoddyl</li>
    <li>python312Packages.example-robot-data</li>
    <li>python312Packages.gepetto-viewer-corba</li>
    <li>python312Packages.gepetto-viewer-corba.doc</li>
    <li>python312Packages.gepetto-viewer-unwrapped</li>
    <li>python312Packages.gepetto-viewer-unwrapped.doc</li>
    <li>python312Packages.omniorb</li>
    <li>python312Packages.omniorbpy</li>
    <li>python312Packages.omniorbpy.dev</li>
    <li>qgv</li>
  </ul>
</details>

nim65s commented:
Result of `nixpkgs-review pr 324018` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>17 packages built:</summary>
  <ul>
    <li>gepetto-viewer</li>
    <li>osgqt</li>
    <li>python311Packages.crocoddyl</li>
    <li>python311Packages.example-robot-data</li>
    <li>python311Packages.omniorb</li>
    <li>python311Packages.omniorbpy</li>
    <li>python311Packages.omniorbpy.dev</li>
    <li>python312Packages.crocoddyl</li>
    <li>python312Packages.example-robot-data</li>
    <li>python312Packages.gepetto-viewer-corba</li>
    <li>python312Packages.gepetto-viewer-corba.doc</li>
    <li>python312Packages.gepetto-viewer-unwrapped</li>
    <li>python312Packages.gepetto-viewer-unwrapped.doc</li>
    <li>python312Packages.omniorb</li>
    <li>python312Packages.omniorbpy</li>
    <li>python312Packages.omniorbpy.dev</li>
    <li>qgv</li>
  </ul>
</details>

nim65s committed:
python-qt: add in python3Packages

nim65s commented:
It is working again on darwin with #333425, so this lgtm now

SuperSandro2000 commented on pkgs/development/python-modules/omniorbpy/default.nix:
```diff
@@ -1,21 +1,24 @@
   fetchurl,
   omniorb,
   pkg-config,
   python3,
```
Using python3* in pythonPackages is strictly forbidden.

SuperSandro2000 commented on pkgs/development/python-modules/omniorbpy/default.nix:
```diff
@@ -2,9 +2,8 @@
-  omniorb,
   pkg-config,
-  python3,
+  python3Packages,
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/python-modules/omniorbpy/default.nix:
```diff
@@ -24,18 +23,18 @@ buildPythonPackage rec {
 
   configureFlags = [
-    "--with-omniorb=${omniorb}"
+    "--with-omniorb=${python3Packages.omniorb}"
```
This was correct before.

SuperSandro2000 commented on pkgs/development/python-modules/omniorbpy/default.nix:
```diff
@@ -24,18 +23,18 @@ buildPythonPackage rec {
-    rm $out/${python3.sitePackages}/omniidl_be/__init__.py
-    rm $out/${python3.sitePackages}/omniidl_be/__pycache__/__init__.*.pyc
+    rm $out/${python3Packages.python.sitePackages}/omniidl_be/__init__.py
+    rm $out/${python3Packages.python.sitePackages}/omniidl_be/__pycache__/__init__.*.pyc
```
Here you want to use python

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -9379,7 +9379,9 @@ self: super: with self; {
-  omniorbpy = callPackage ../development/python-modules/omniorbpy { };
+  omniorb = toPythonModule (pkgs.omniorb.override { python3 = self.python; });
+
+  omniorbpy = callPackage ../development/python-modules/omniorbpy { python3Packages = self; };
```
We also don't need this when not using python3*

SuperSandro2000 commented on pkgs/development/python-modules/omniorbpy/default.nix:
Please collapse with the last commit(s)

SuperSandro2000 commented on pkgs/development/tools/omniorb/default.nix:
```diff
@@ -16,24 +16,24 @@ stdenv.mkDerivation rec {
 
   nativeBuildInputs = [ pkg-config ];
-  buildInputs = [ python3 ];
+  buildInputs = [ python3Packages.python ];
```
You can use python3.pkgs here to keep the diff smaller.

SuperSandro2000 commented on pkgs/by-name/qg/qgv/package.nix:
```diff
@@ -0,0 +1,43 @@
+
+  src = fetchFromGitHub {
+    owner = "gepetto";
+    repo = finalAttrs.pname;
```
```suggestion
    repo = "qgv";
```

SuperSandro2000 commented on pkgs/development/python-modules/gepetto-viewer-unwrapped/default.nix:
```diff
@@ -0,0 +1,71 @@
+  openscenegraph,
+  osgqt,
+  pkg-config,
+  python3Packages,
```
```suggestion
```

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -4901,6 +4901,10 @@ self: super: with self; {
 
+  gepetto-viewer-corba = callPackage ../development/python-modules/gepetto-viewer-corba { };
+
+  gepetto-viewer-unwrapped = callPackage ../development/python-modules/gepetto-viewer-unwrapped { python3Packages = self; };
```
This should be without -unwrapped

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -4901,6 +4901,10 @@ self: super: with self; {
 
   georss-wa-dfes-client = callPackage ../development/python-modules/georss-wa-dfes-client { };
 
+  gepetto-viewer-corba = callPackage ../development/python-modules/gepetto-viewer-corba { };
```
This is in the wrong commit

SuperSandro2000 commented on pkgs/development/python-modules/gepetto-viewer-corba/default.nix:
```diff
@@ -0,0 +1,61 @@
+  doCheck = true;
+
+  meta = {
+    homepage = "https://github.com/gepetto/gepetto-viewer-corba";
```
Missing description

SuperSandro2000 commented on pkgs/development/python-modules/gepetto-viewer-corba/default.nix:
```diff
@@ -0,0 +1,61 @@
+  doxygen,
+  omniorb,
+  pkg-config,
+  python3Packages,
```
```suggestion
```

SuperSandro2000 commented on pkgs/by-name/ge/gepetto-viewer/package.nix:
```diff
@@ -0,0 +1,22 @@
+  stdenv,
+}:
+stdenv.mkDerivation {
+  inherit (python3Packages.gepetto-viewer-unwrapped) pname version meta;
```
I don't understand why we don't combine this and the unwrapped package

nim65s commented:
@SuperSandro2000 : thanks for the review ! It makes more sense now.

nim65s commented:
Oh, wait, there is an easy solution :)
I updated the example one more time.

nim65s commented:
Result of `nixpkgs-review pr 324018` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>24 packages built:</summary>
  <ul>
    <li>gepetto-viewer</li>
    <li>gepetto-viewer.doc</li>
    <li>osgqt</li>
    <li>python-qt (python312Packages.python-qt)</li>
    <li>python311Packages.crocoddyl</li>
    <li>python311Packages.example-robot-data</li>
    <li>python311Packages.gepetto-viewer</li>
    <li>python311Packages.gepetto-viewer-corba</li>
    <li>python311Packages.gepetto-viewer-corba.doc</li>
    <li>python311Packages.gepetto-viewer.doc</li>
    <li>python311Packages.omniorb</li>
    <li>python311Packages.omniorbpy</li>
    <li>python311Packages.omniorbpy.dev</li>
    <li>python311Packages.python-qt</li>
    <li>python312Packages.crocoddyl</li>
    <li>python312Packages.example-robot-data</li>
    <li>python312Packages.gepetto-viewer</li>
    <li>python312Packages.gepetto-viewer-corba</li>
    <li>python312Packages.gepetto-viewer-corba.doc</li>
    <li>python312Packages.gepetto-viewer.doc</li>
    <li>python312Packages.omniorb</li>
    <li>python312Packages.omniorbpy</li>
    <li>python312Packages.omniorbpy.dev</li>
    <li>qgv</li>
  </ul>
</details>

nim65s commented:
Result of `nixpkgs-review pr 324018` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>24 packages built:</summary>
  <ul>
    <li>gepetto-viewer</li>
    <li>gepetto-viewer.doc</li>
    <li>osgqt</li>
    <li>python311Packages.crocoddyl</li>
    <li>python311Packages.example-robot-data</li>
    <li>python311Packages.gepetto-viewer</li>
    <li>python311Packages.gepetto-viewer-corba</li>
    <li>python311Packages.gepetto-viewer-corba.doc</li>
    <li>python311Packages.gepetto-viewer.doc</li>
    <li>python311Packages.omniorb</li>
    <li>python311Packages.omniorbpy</li>
    <li>python311Packages.omniorbpy.dev</li>
    <li>python311Packages.python-qt</li>
    <li>python312Packages.crocoddyl</li>
    <li>python312Packages.example-robot-data</li>
    <li>python312Packages.gepetto-viewer</li>
    <li>python312Packages.gepetto-viewer-corba</li>
    <li>python312Packages.gepetto-viewer-corba.doc</li>
    <li>python312Packages.gepetto-viewer.doc</li>
    <li>python312Packages.omniorb</li>
    <li>python312Packages.omniorbpy</li>
    <li>python312Packages.omniorbpy.dev</li>
    <li>python312Packages.python-qt</li>
    <li>qgv</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/development/python-modules/gepetto-viewer/default.nix:
```diff
@@ -9,65 +9,83 @@
-  preFixup = lib.optionalString stdenv.isDarwin "export LC_ALL=C";
+    passthru.withPlugins =
+      plugins:
+      stdenv.mkDerivation {
```
Why not use runCommand?

nim65s replied:
The OP

nim65s replied:
I didn't knew how to use it, but yes, it seems way better, thanks !

nim65s commented on pkgs/development/libraries/osgqt/package.nix:
```diff
@@ -0,0 +1,49 @@
+  ];
+
+  postPatch = ''
+    substituteInPlace CMakeLists.txt --replace \
+      "FIND_PACKAGE(Qt5Widgets REQUIRED)" \
+      "FIND_PACKAGE(Qt5Widgets REQUIRED)
+       FIND_PACKAGE(Qt5OpenGL REQUIRED)"
```
good catch, thanks

nim65s commented on pkgs/development/python-modules/gepetto-viewer/default.nix:
```diff
@@ -0,0 +1,73 @@
+  doCheck = true;
+
+  # wrapQtAppsHook uses isMachO, which fails to detect binaries without this
+  # ref. https://github.com/NixOS/nixpkgs/pull/138334
```
which itself ref https://github.com/NixOS/nixpkgs/pull/133808 OP:
> But it needs LC_ALL=C in order to work, otherwise read -n 4 would not read 4 bytes, and I am not sure why. I couldn't find information about what stdenv does regarding locale.

nim65s committed:
gepetto-viewer-corba: init at 5.8.0
